src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http3WebtransportManager.cs

```diff
@@ -0,0 +1,120 @@
+    [SupportedOSPlatform("macos")]
+    internal sealed class Http3WebtransportManager : IDisposable, IAsyncDisposable
+    {
+        private ConcurrentDictionary<long, Http3WebtransportSession> _sessions;
+        private QuicConnection _connection;
+
+        public Http3WebtransportManager(QuicConnection connection)
+        {
+            _sessions = new ConcurrentDictionary<long, Http3WebtransportSession>();
+            _connection = connection;
+        }
+
+        public bool AddSession(QuicStream connectStream, Http3WebtransportSession webtransportSession)
+        {
+            bool ans = _sessions.TryAdd(connectStream.Id, webtransportSession);
+            return ans;
+        }
+
+        public bool FindSession(long streamId, out Http3WebtransportSession? session)
+        {
+            return _sessions.TryGetValue(streamId, out session);
+        }
+
+        public void AcceptServerStream(QuicStream stream, long sessionId)
+        {
+            Http3WebtransportSession? session;
+            _sessions.TryGetValue(sessionId, out session);
+            // if no session with that id exists throw exception
+            // https://datatracker.ietf.org/doc/html/draft-ietf-webtrans-http3#section-4
+            if (session == null)
+            {
+                stream.Abort(QuicAbortDirection.Both, (long)Http3ErrorCode.WebtransportBufferedStreamRejected);
+                return;
+            }
+            session.AcceptServerStream(stream);
+        }
+
+        public async ValueTask<QuicStream?> CreateClientStream(QuicStreamType type, long sessionId)
+        {
+            QuicStream clientWebtransportStream;
+            try
+            {
+                clientWebtransportStream = await _connection.OpenOutboundStreamAsync(type).ConfigureAwait(false);
+                await clientWebtransportStream.WriteAsync(BuildWebtransportStreamClientFrame(type, sessionId), default).ConfigureAwait(false);
+
+                return clientWebtransportStream;
+            }
+            catch (Exception ex)
+            {
+                throw ex;
+            }
+        }
+
+        private byte[] BuildWebtransportStreamClientFrame(QuicStreamType type, long sessionId)
+        {
+            Span<byte> buffer = stackalloc byte[2 + VariableLengthIntegerHelper.MaximumEncodedLength];
+            int webtransportLength;
+            if (type == QuicStreamType.Unidirectional)
+                webtransportLength = VariableLengthIntegerHelper.WriteInteger(buffer.Slice(0), (long)Http3StreamType.WebTransportUnidirectional);
+            else
+                webtransportLength = VariableLengthIntegerHelper.WriteInteger(buffer.Slice(0), (long)Http3StreamType.WebTransportBidirectional);
+            int webtransportSessionLength = VariableLengthIntegerHelper.WriteInteger(buffer.Slice(2), (long)sessionId);
+            int payloadLength = webtransportLength + webtransportSessionLength; // includes the webtransport stream and the session id
```

Comment:
In the code above, you are mixing up calculated length of stream type and hardcoded length of stream type 🙂